test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs

```diff
@@ -252,6 +252,304 @@ public static TheoryData MultiAttributeCheckBoxData
             Assert.Equal(expectedPostElement, output.PostElement.GetContent());
         }
 
+        [Theory]
```

Comment:
Dislike using theory here. These are each truly different tests since their call paths totally diverge inside the `InputTagHelper`. Make a helper method that does all the below magic and takes in `inputTypeName` but have separate `[Fact]`s for each of these.

Reply:
Actually, you didn't touch any TagHelper specific control flow code that throws within the TagHelper. Shouldn't all these "Throw" tests be in the HtmlGenerator specific tests as unit tests? Just asking because all these TagHelper integration tests are super heavy.

Reply:
1. Distinct tests make sense when the scenarios are truly different, especially if the combination leads to conditions (worse conditional assertions) in the test. In this case, distinct tests would look like
  ``` c#
  [Fact]
  public void Process_Checkbox_WithEmptyForName_Throws()
  {
      Process_WithEmptyForName_Throws("checkbox");
  }

  // ...

  public void Process_WithEmptyForName_Throws(string inputTypeName)
  {
      // Exactly the code that's in there now.
  }
  ```
2. Testing `DefaultHtmlGenerator` isn't done much at all and I'm not going to change that in this PR. The main issue is the very different ways string-based HTML helpers, expression-based HTML helpers, and tag helpers use the methods. But, it's also difficult to ensure the scenario (e.g. `<input type="checkbox" ... />` `throw`s when `fullName` would end up empty) is actually tested. We'd end up with a tag helper test using a `Mock<IHtmlGenerator>` making sure the expected data is passed and a `DefaultHtmlGenerator` test passing the same data to the method mocked in the first test. In this particular case, I'm also keen to keep the "...Throws..." and "...DoesNotThrow..." tests together, making it easy to see the small change that avoids the `Exception` and to maintain the tests (if that's necessary in the future).